File: simulator/Organisms/Animals/Human.py
from simulator.Organisms.Animal import Animal
import pygame

pygame.init()


class Human(Animal):
    def __init__(self, current_world, x_coordinate, y_coordinate):
        super().__init__(current_world, x_coordinate, y_coordinate, 5, 4, True, 'X', 'simulator\\images\\human.png')
        self.__shield_duration = 0
        self.__shield_cooldown = 0

    def shield_active(self):
        if self.get_shield_dur() == 0:
            return False
        return True

    def on_cooldown(self):
        if self.get_shield_cooldown() == 0:
            return False
        return True

    def get_shield_dur(self):
        return self.__shield_duration

    def get_shield_cooldown(self):
        return self.__shield_cooldown

    def set_duration(self, duration):
        self.__shield_duration = duration

    def set_cooldown(self, cooldown):
        self.__shield_cooldown = cooldown

    def action(self):
        if self.shield_active():
            self.set_duration(self.get_shield_dur() - 1)
            if self.get_shield_dur() == 0:
                text = "Alzurs Shield wears off"
                self.set_image('simulator\\images\\human.png')
                self.get_world().add_log(text)
        if self.on_cooldown():
            self.set_cooldown(self.get_shield_cooldown() - 1)
        new_x = self.get_x()
        new_y = self.get_y()
        legal_move = False
        ability_pressed = False
        while not (legal_move or ability_pressed):
            for event in pygame.event.get():
                if event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_LEFT:
                        if not self.get_world().is_outside(new_x - 1, new_y):
                            new_x -= 1
                            legal_move = True
                    elif event.key == pygame.K_RIGHT:
                        if not self.get_world().is_outside(new_x + 1, new_y):
                            new_x += 1
                            legal_move = True
                    elif event.key == pygame.K_UP:
                        if not self.get_world().is_outside(new_x, new_y - 1):
                            new_y -= 1
                            legal_move = True
                    elif event.key == pygame.K_DOWN:
                        if not self.get_world().is_outside(new_x, new_y + 1):
                            new_y += 1
                            legal_move = True
                    elif event.key == pygame.K_SPACE and not self.on_cooldown():
                        self.set_duration(5)
                        self.set_cooldown(10)
                        text = "Alzurs shield activated"
                        self.set_image('simulator\\images\\human_shield.png')
                        self.get_world().add_log(text)
                        ability_pressed = True
                    elif event.key == pygame.K_s:
                        self.get_world().save()
                elif event.type == pygame.QUIT:
                    pygame.quit()
                    quit()
        if not ability_pressed:
            if self.get_world().is_occupied(new_x, new_y):
                self.get_world().get_field(new_x, new_y).collision(self)
            else:
                self.new_location(new_x, new_y)

    def collision(self, predator):
        text = str(predator.get_token()) + " attacks " + str(self.get_token()) + " "
        if self.shield_active():
            text += "X defends"
            self.get_world().add_log(text)
            self.get_world().del_from_map(predator)
            predator.set_position(self.get_x(), self.get_y())
            predator.move()
            self.get_world().set_field(self.get_x(), self.get_y(), self)
        else:
            if (self.get_strength() > predator.get_strength()
                    or (self.get_strength() == predator.get_strength() and self.get_age() > predator.get_age())):
                text += str(self.get_token()) + " wins"
                self.get_world().add_log(text)
                self.get_world().del_organism(predator)
            else:
                text += str(predator.get_token()) + " wins"
                self.get_world().add_log(text)
                tmp_x = self.get_x()
                tmp_y = self.get_y()
                self.get_world().del_organism(self)
                predator.new_location(tmp_x, tmp_y)

File: world_simulator_cpp/Human.cpp
#include <iostream>
#include <conio.h>
#include "World.h"
#include "Human.h"
#include "enumList.h"
using namespace std;

Human::Human(World* currentWorld, int X_coordinate, int Y_coordinate) : Animal(currentWorld, X_coordinate, Y_coordinate, 5, 4, 'X') {
	this->GetWorld()->SetHumanAlive(true);
}

bool Human::ShieldActive() {
	if (this->ShieldDuration == 0)
		return false;
	return true;
}
bool Human::OnCooldown() {
	if (this->ShieldCooldown == 0)
		return false;
	return true;
}
int Human::GetShieldDuration() {
	return this->ShieldDuration;
}
int Human::GetShieldCooldown() {
	return this->ShieldCooldown;
}

void Human::SetShieldDuration(int duration) {
	this->ShieldDuration = duration;
}
void Human::SetShieldCooldown(int cooldown) {
	this->ShieldCooldown = cooldown;
}

void Human::Action() {
	if (this->ShieldActive()) {
		this->SetShieldDuration(this->GetShieldDuration() - 1);
	}
	if (this->OnCooldown()) {
		this->SetShieldCooldown(this->GetShieldCooldown() - 1);
	}
	char move;
	bool goodKey = false;
	int new_x = this->GetX(), new_y = this->GetY();
	while (!goodKey) {
		move = _getch();
		if (move == 72 || move == 75 || move == 77 || move == 80) {
			goodKey = true;
			if (move == 72) {
				if (this->GetWorld()->IsOutsideBorder(this->GetX(), this->GetY() - 1))
					goodKey = false;
				else
					new_y--;
			}
			else if (move == 75) {
				if (this->GetWorld()->IsOutsideBorder(this->GetX() - 1, this->GetY()))
					goodKey = false;
				else
					new_x--;
			}
			else if (move == 80) {
				if (this->GetWorld()->IsOutsideBorder(this->GetX(), GetY() + 1))
					goodKey = false;
				else
					new_y++;
			}
			else if (move == 77) {
				if (this->GetWorld()->IsOutsideBorder(this->GetX() + 1, GetY()))
					goodKey = false;
				else
					new_x++;
			}
			if (goodKey == false)
				cout << "You can't move there, try different direction" << endl;
		}
		else if (move == 32) {
			if (this->OnCooldown()) {
				cout << "You can't use the shield for another " << this->GetShieldCooldown() << " rounds" << endl;
			}
			else {
				this->SetShieldDuration(5);
				this->SetShieldCooldown(10);
				cout << "Alzur's Shield activated" << endl;
			}
		}	
	}
	if (this->currentWorld->IsFieldOccupied(new_x, new_y)) {
		this->currentWorld->GetField(new_x, new_y)->Collision(this, this->currentWorld->GetField(new_x, new_y));
	}
	else {
		ChangeLocation(new_x, new_y);
		cout << this->GetToken() << " moves to x: " << new_x << " y: " << new_y << endl;
	}
}

void Human::Collision(Organism* predator, Organism* prey) {
	if (ShieldActive()) {
		bool LegalMove = false;
		int new_x, new_y;
		while (!LegalMove) {
			new_x = this->GetX();
			new_y = this->GetY();
			int direction = rand() % 4;
			if (direction == DOWN) {
				new_y++;
				if (!currentWorld->IsOutsideBorder(new_x, new_y))
					LegalMove = true;
			}
			else if (direction == RIGHT) {
				new_x++;
				if (!currentWorld->IsOutsideBorder(new_x, new_y))
					LegalMove = true;
			}
			else if (direction == UP) {
				new_y--;
				if (!currentWorld->IsOutsideBorder(new_x, new_y))
					LegalMove = true;
			}
			else {
				new_x--;
				if (!currentWorld->IsOutsideBorder(new_x, new_y))
					LegalMove = true;
			}
			if (LegalMove) {
				if (this->currentWorld->IsFieldOccupied(new_x, new_y)) {
					this->currentWorld->GetField(new_x, new_y)->Collision(predator, this->currentWorld->GetField(new_x, new_y));
				}
				else {
					predator->ChangeLocation(new_x, new_y);
					predator->Draw();
				}
			}
		}
		cout << this->GetToken() << " defends attack from " << predator->GetToken() << ". " << predator->GetToken() << " moves to x: " << predator->GetX() << " y: " << predator->GetY() << endl;
	}
	else {
		cout << predator->GetToken() << " attacks X on x: " << prey->GetX() << " y: " << prey->GetY() << endl;
		if (prey->GetStrength() > predator->GetStrength()) {
			this->GetWorld()->DelFromWorldMap(predator);
			this->GetWorld()->DelFromOrganismList(predator);
			delete predator;
		}
		else {
			this->GetWorld()->SetHumanAlive(false);
			predator->ChangeLocation(prey->GetX(), prey->GetY());
			this->GetWorld()->DelFromWorldMap(prey);
			this->GetWorld()->DelFromOrganismList(prey);
			predator->Draw();
		}
	}
}

Human::~Human() {
	currentWorld->SetHumanAlive(false);
}